storage/remote/queue_manager.go

```diff
@@ -440,36 +443,50 @@ func (t *QueueManager) calculateDesiredShards() {
 	// (received - send) so we can catch up with any backlog. We use the average
 	// outgoing batch latency to work out how many shards we need.
 	var (
-		samplesIn          = t.samplesIn.rate()
-		samplesOut         = t.samplesOut.rate()
-		samplesKeptRatio   = samplesOut / (t.samplesDropped.rate() + samplesOut)
-		samplesOutDuration = t.samplesOutDuration.rate()
+		samplesInRate      = t.samplesIn.rate()
+		samplesOutRate     = t.samplesOut.rate()
+		samplesKeptRatio   = samplesOutRate / (t.samplesDropped.rate() + samplesOutRate)
+		samplesOutDuration = t.samplesOutDuration.rate() / float64(time.Second)
+		samplesPendingRate = samplesInRate*samplesKeptRatio - samplesOutRate
 		highestSent        = t.highestSentTimestampMetric.Get()
 		highestRecv        = highestTimestamp.Get()
-		samplesPending     = (highestRecv - highestSent) * samplesIn * samplesKeptRatio
+		samplesPending     = (highestRecv - highestSent) * samplesInRate * samplesKeptRatio
 	)
 
-	// We use an integral accumulator, like in a PID, to help dampen oscillation.
-	t.integralAccumulator = t.integralAccumulator + (samplesPending * 0.1)
-
-	if samplesOut <= 0 {
+	if samplesOutRate <= 0 {
```

Comment:
Actually I find some problems here :joy: 

If the network between prometheus and remote storage is shaking and shard number become 1000 after reshard.

However the network traffic of 1000 shard is a disaster for remote storage which can't be suffered and it crushed.

Then remote storage will restart again and again, because the sampleOutRate is 0, and the number of shard still be 1000 which result a deadlock.

So if the samplesOutRate is 0, maybe we should return shard number back to min to make the network recover.

Reply:
That's an interesting problem, if you are backed up far enough then most users would want to catch up as quickly as possible, so scaling to max shards makes sense to me. If the remote storage cannot keep up, it is likely you will want a smaller `max_shards` that the remote storage can handle. Curious what others think as well.

Reply:
I have always thought 1,000 is way too high a default. However people demanded proof.

Reply:
We should ensure the default can handle a 1M/s instance, with slack for catchup.

Reply:
I don't know if there's really a nice way that resharding can account for issues on the remote storage's end. Maybe we could take into account retries somehow, but reconfiguring Prometheus and the remote storage is going to yield a better result. 

If the amount of network traffic and sheer number of requests is too much for the remote storage to handle you can either scale up your remote storage or mess with your remote write config.

Reply:
Yeah, there are strong reasons for 1k shards (which can each do 1k samples/s), and we've made the limit configurable to support other usecases.

I don't wonder wether we can improve the backoff so that we only ever for 10 QPS per shard, to prevent us DoSing the endpoint when it is down... but thats a separate issue.